### PR TITLE
feat: Cart Customer link + create cart with customer

### PR DESCRIPTION
**Pull request by olivermrbl**
**What**
- Add Cart Customer read-only link
- Add `findOrCreateCustomer` step to `CreateCartWorkflow`. Three scenarios are covered:
  - Retrieve customer from passed ID
  - Retrieve customer from passed email
  - Create customer with passed email
- Use signed-in customer id in `POST /store/carts`
- Restructure tests

**Tests**
- Cart Customer link queries
- Create cart with existing customer
- Create cart creates customer using email
- Compensate: Should delete customer, if created from email
- Compensate: Should not delete customer, if customer already existed

_Note: Compensation flows are tested using the local workflow instead of the API, as we cannot modify the global workflow (yet)_

### CODE REVIEW

File: packages/core-flows/src/definition/cart/steps/find-or-create-customer.ts [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+
+export interface StepOutput {
+  customer?: any
+  email?: string
+}
+
+interface StepCompensateInput {
+  customer?: any
+  customerWasCreated: boolean
+}
```
Comment (olivermrbl) [inline packages/core-flows/src/definition/cart/steps/find-or-create-customer.ts]: comment: TODO

File: integration-tests/plugins/__tests__/cart/store/carts.spec.ts [hunk 2]
```diff
@@ -41,167 +50,331 @@ describe("POST /store/carts", () => {
...
+      const created = await api.post(`/store/carts`, {
         currency_code: "usd",
-        email: "tony@stark.com",
-        region: expect.objectContaining({
-          id: region.id,
-          currency_code: "usd",
-        }),
+        region_id: region.id,
         sales_channel_id: salesChannel.id,
+        customer_id: customer.id,
```
Comment (srindom) [inline integration-tests/plugins/__tests__/cart/store/carts.spec.ts]: thought: I don't think we should allow passing the customer id directly form the client. Either you get a guest customer or you use the logged in customer's id. Otherwise, you would be able to impersonate others.

File: integration-tests/plugins/__tests__/cart/store/carts.spec.ts [hunk 3]
```diff
@@ -41,167 +50,331 @@ describe("POST /store/carts", () => {
...
+          "throw",
+          findOrCreateCustomerStepId,
+          {
+            invoke: async function failStep() {
+              throw new Error(`Failed to create cart`)
+            },
+          },
+          {
+            noCompensation: true,
+          }
```
Comment (srindom) [inline integration-tests/plugins/__tests__/cart/store/carts.spec.ts]: q: why is this necessary? Doesn't the step just fail and only previous steps compensate?

File: packages/medusa/src/api-v2/store/carts/route.ts [hunk 4]
```diff
@@ -1,14 +1,19 @@
...
+import { StorePostCartsCartReq } from "./validators"
 
 export const POST = async (req: MedusaRequest, res: MedusaResponse) => {
-  const workflow = createCartWorkflow(req.scope)
+  const input = req.validatedBody as StorePostCartsCartReq
 
-  const { result, errors } = await workflow.run({
-    input: req.validatedBody as CreateCartDTO,
+  // If the customer is logged in, we auto-assign them to the cart
+  if (req?.user?.customer_id) {
```
Comment (srindom) [inline packages/medusa/src/api-v2/store/carts/route.ts]: thought: not sure if we made changes but the customer id from the authuser has been dug out like so in the /me route:

https://github.com/medusajs/medusa/blob/develop/packages/medusa/src/api-v2/store/customers/me/route.ts

File: integration-tests/plugins/__tests__/cart/store/carts.spec.ts (hunk 2 above)
Comment (olivermrbl) [reply integration-tests/plugins/__tests__/cart/store/carts.spec.ts]: Do you mean they will be able to impersonate others by "guessing" their customer ID? Many other platforms' Cart APIs allow customer ID. Are you sure we should remove it?

We've also always had `customer_id` exposed in the update cart endpoint.

File: packages/medusa/src/api-v2/store/carts/route.ts (hunk 4 above)
Comment (olivermrbl) [reply packages/medusa/src/api-v2/store/carts/route.ts]: Nice catch, I forgot to add in the new middleware.

File: integration-tests/plugins/__tests__/cart/store/carts.spec.ts (hunk 2 above)
Comment (srindom) [reply integration-tests/plugins/__tests__/cart/store/carts.spec.ts]: My thinking is that if someone has an account then that person must be logged in, in order to have their customer id associated with the cart. 

Guessing is close to being mathematically impossible but if you somehow have your customer id exposed it should still be protected by a login to be used. 

For guest customers it is less important but I don't really see a need for a customer id being editable from outside in regular use cases (i.e., our default apis scope).

Comment (olivermrbl) [reply integration-tests/plugins/__tests__/cart/store/carts.spec.ts]: > Guessing is close to being mathematically impossible

Yeah, this is why I was questioning it, but I think your other points make sense. Will remove it.

File: integration-tests/plugins/__tests__/cart/store/carts.spec.ts (hunk 3 above)
Comment (olivermrbl) [reply integration-tests/plugins/__tests__/cart/store/carts.spec.ts]: Made a c/p. Think this can def. be removed.

**Review by srindom — APPROVED** [review]
LGTM!

File: packages/medusa/src/api-v2/store/carts/middlewares.ts [hunk 5]
```diff
@@ -8,6 +9,15 @@ import {
 } from "./validators"
 
 export const storeCartRoutesMiddlewares: MiddlewareRoute[] = [
+  {
+    method: "ALL",
```
Comment (olivermrbl) [inline packages/medusa/src/api-v2/store/carts/middlewares.ts:13]: @pKorsholm, can I get you to verify the setup here is correct?
